tokio-net/src/driver/reactor/mod.rs

```diff
@@ -434,31 +426,16 @@ impl Inner {
     ///
     /// The registration token is returned.
     pub(super) fn add_source(&self, source: &dyn Evented) -> io::Result<usize> {
-        // Get an ABA guard value
         let aba_guard = self.next_aba_guard.fetch_add(1 << TOKEN_SHIFT, Relaxed);
-
-        let key = {
-            // Block to contain the write lock
-            let mut io_dispatch = self.io_dispatch.write();
-
-            if io_dispatch.len() == MAX_SOURCES {
-                return Err(io::Error::new(
-                    io::ErrorKind::Other,
-                    "reactor at max \
-                     registered I/O resources",
-                ));
-            }
-
-            io_dispatch.insert(ScheduledIo {
-                aba_guard,
-                readiness: AtomicUsize::new(0),
-                reader: AtomicWaker::new(),
-                writer: AtomicWaker::new(),
-            })
-        };
-
-        let token = aba_guard | key;
-        debug!(message = "adding I/O source", token);
+        let index = self.io_dispatch.insert(aba_guard).ok_or_else(|| {
```

Comment:
I know this is what we did before. However, what do you think about instead of having a "global" ABA guard, we have a per slot ABA guard. So, you already have an `aba_guard` field on `ScheduledIo`, so "just" use that value for the aba guard? Then, on remove, it is incremented. That should reduce the likelihood of false matches.